cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/partitioned/PartitionedFileSetDataset.java

```diff
@@ -113,10 +117,41 @@ public Partitioning getPartitioning() {
 
   @Override
   public void addPartition(PartitionKey key, String path) {
-    addPartition(key, path, true);
+    addPartition(key, path, new PartitionMetadata(ImmutableMap.<String, String>of()));
   }
 
-  protected void addPartition(PartitionKey key, String path, boolean addToExplore) {
+  @Override
+  public void addPartition(PartitionKey key, String path, PartitionMetadata metadata) {
+    addPartition(key, path, true, metadata);
+  }
+
+  @Override
+  public void updateMetadata(PartitionKey key, PartitionMetadata metadata) {
+    final byte[] rowKey = generateRowKey(key, partitioning);
+    Row row = partitionsTable.get(rowKey);
+    if (row == null) {
+      throw new DataSetException(String.format("Dataset '%s' does not have a partition for key: %s", getName(), key));
+    }
+
+    // delete all the old metadata and then write the new metadata
+    Delete delete = new Delete(rowKey);
+    for (Map.Entry<byte[], byte[]> entry : row.getColumns().entrySet()) {
```

Comment:
To be consistent with the usage of "addMetadataToPut" method, I think this for loop should also be abstracted out in a method.